src/main/kotlin/brs/services/impl/BlockServiceImpl.kt

```diff
@@ -79,11 +81,27 @@ class BlockServiceImpl(private val dp: DependencyProvider) : BlockService {
                 logger.safeDebug { "Block at height ${block.height} was not pre-verified! Pre-verification threads are probably not keeping up..." }
             }
 
+            val checkPointHeight = dp.propertyService.get(if (dp.propertyService.get(Props.DEV_TESTNET)) Props.DEV_CHECKPOINT_HEIGHT else Props.BRS_CHECKPOINT_HEIGHT)
             val pocTime = try {
-                if (scoopData == null) {
+                if(block.height < checkPointHeight){
+                    // do not verify the nonce up to the checkpoint block (lowers CPU usage)
+                    BigInteger.valueOf(0L)
+                }
+                else {
+                    if(block.height == checkPointHeight) {
+                        val checkPointHash =
+                            dp.propertyService.get(if (dp.propertyService.get(Props.DEV_TESTNET)) Props.DEV_CHECKPOINT_HASH else Props.BRS_CHECKPOINT_HASH)
+
+                        val receivedHash = Hex.toHexString(block.previousBlockHash)
+                        if (!receivedHash.equals(checkPointHash)) {
+                            throw BlockchainProcessorService.BlockNotAcceptedException("Error pre-verifying checkpoint block hash")
```

Comment:
What happens at this point? You have an incorrect blockchain (which occurs surprisingly often) and no way of finding how far to roll back - the only option is to fully reset the DB

Reply:
What about an exponential back-off? i.e. pop-off 5 blocks, then 25, then 125, etc

Reply:
I don't think you can have an incorrect chain thousands of blocks in the past. This can be even seen as an extra security check.